src/metabase/lib/js.cljs

```diff
@@ -117,28 +119,81 @@
+;; `display-info` is called multiple times on the same opaque CLJS value, it will be cached.
+;;
+;; [[display-info*]] is the inner implementation. It calls [[lib.core/display-info]] to get the CLJS form, then
+;; [[display-info->js]] to convert it to JS.
+;;
+;; JS conversion in the tricky cases (maps and seqs) are handled by separate, LRU-cached functions
+;; [[display-info-map->js]] and [[display-info-seq->js]]. Keywords are converted with [[u/qualified-name]].
+;;
+;; [[display-info-map->js]] converts CLJS maps to JS objects. Keys are converted from `:kebab-case-keywords` to
+;; `"camelCaseStrings"`. Values are recursively converted by [[display-info->js]]. (Note that this passes through the
+;; LRU caches for nested maps and seqs - this is important since many inner pieces are reused across eg. columns.)
+
+;; [[display-info-seq->js]] converts CLJS `sequential?` things to JS arrays, recursively calling [[display-info->js]] on
+;; each element. (This is cached just like map values above.)
+
+;; **Note:** there's an important property here that's worth calling out explicitly. Two columns which are `=` can
+;; be returned by eg. `visible-columns` on two different queries. This is why it's important that the caches used where
+;; `a-query` and `stage-number` are provided ([[display-info]]) are side channel caches that only cache for a specific
+;; instance, not for any `=` value. The different queries might lead to different display names or other details.
+;; The CLJS -> JS conversion doesn't know about queries, so it uses `=`-based LRU caches.
+(declare ^:private display-info->js)
+
+(defn- display-info-map->js* [x]
+  (reduce (fn [obj [cljs-key cljs-val]]
+            (set! js/window.__DIJS_recurses (inc js/window.__DIJS_recurses))
+            (let [js-key (-> cljs-key u/qualified-name u/->camelCaseEn)
+                  js-val (display-info->js cljs-val)] ;; Recursing through the cache
+              (gobject/set obj js-key js-val)
+              obj))
+          (js-obj)
+          x))
+
+(def ^:private display-info-map->js
+  (memoize/lru display-info-map->js* :lru/threshold 256))
+
+(defn- display-info-seq->js* [x]
+  (to-array (map display-info->js x)))
+
+(def ^:private display-info-seq->js
+  (memoize/lru display-info-seq->js* :lru/threshold 256))
+
+(defn- display-info->js
+  "Converts CLJS [[lib.core/display-info]] results into JS objects for the FE to consume.
+  Recursively converts CLJS maps and `sequential?` things likewise."
   [x]
   (cond
-    (map? x)        (-> x
-                        (update-keys u/->camelCaseEn)
-                        (update-vals display-info-js))
-    (sequential? x) (map display-info-js x)
+    ;; Note that map? is only true for CLJS maps, not JS objects.
+    (map? x)        (display-info-map->js x)
+    ;; Likewise, JS arrays are not sequential? while CLJS vectors, seqs and sets are.
+    (sequential? x) (display-info-seq->js x)
+    (keyword? x)    (u/qualified-name x)
```

Comment:
It feels asymmetrical that we don't camelCase here.

Reply:
Perhaps, but this is only for the "value" side, not for map keys. This is the original behavior: only map keys get camelcased. We leave keyword values like `:type/Number` alone - actually these values are already mostly camelcased anyway.